File: app/src/main/java/com/example/earthwallet/ui/pages/anml/ANMLClaimMainFragment.java
package com.example.earthwallet.ui.pages.anml;

import com.example.earthwallet.R;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;

import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;
 
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.appcompat.app.AlertDialog;
import androidx.security.crypto.EncryptedSharedPreferences;
import androidx.security.crypto.MasterKeys;

import com.example.earthwallet.bridge.activities.TransactionActivity;
import com.example.earthwallet.wallet.services.SecretWallet;
import com.example.earthwallet.Constants;
import com.example.earthwallet.ui.components.LoadingOverlay;
import com.example.earthwallet.bridge.services.SecretQueryService;
import com.example.earthwallet.ui.pages.anml.ANMLRegisterFragment;
import com.example.earthwallet.ui.pages.anml.ANMLClaimFragment;
import com.example.earthwallet.ui.pages.anml.ANMLCompleteFragment;

import org.json.JSONObject;

public class ANMLClaimMainFragment extends Fragment implements ANMLRegisterFragment.ANMLRegisterListener, ANMLClaimFragment.ANMLClaimListener {
    private static final String TAG = "ANMLClaimFragment";
    private static final String PREF_FILE = "secret_wallet_prefs";

    private static final long ONE_DAY_MILLIS = 24L * 60L * 60L * 1000L;

    private TextView errorText;
    private LoadingOverlay loadingOverlay;
    private View fragmentContainer;
    
    private SharedPreferences securePrefs;
    private boolean suppressNextQueryDialog = false;
    private String currentRegistrationReward;

    // Request codes for launching bridge Activities
    private static final int REQ_QUERY = 1001;
    private static final int REQ_EXECUTE = 1002;

    // Broadcast receiver for transaction success
    private BroadcastReceiver transactionSuccessReceiver;

    public ANMLClaimMainFragment() {}
    
    public static ANMLClaimMainFragment newInstance() {
        return new ANMLClaimMainFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_anml_claim_main, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        try {
            SecretWallet.initialize(getContext());
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize SecretWallet", e);
        }

        errorText = view.findViewById(R.id.anml_error_text);
        loadingOverlay = view.findViewById(R.id.loading_overlay);
        fragmentContainer = view.findViewById(R.id.anml_root);

        // Initialize the loading overlay with this fragment for Glide
        if (loadingOverlay != null) {
            loadingOverlay.initializeWithFragment(this);
        }

        initSecurePrefs();
        setupBroadcastReceiver();
        registerBroadcastReceiver();

        // Start status check
        checkStatus();
    }

    private void initSecurePrefs() {
        // Use centralized secure preferences from HostActivity
        securePrefs = ((com.example.earthwallet.ui.host.HostActivity) getActivity()).getSecurePrefs();
    }

    private void setupBroadcastReceiver() {
        transactionSuccessReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                Log.d(TAG, "Received TRANSACTION_SUCCESS broadcast - refreshing ANML status immediately");

                // Start multiple refresh attempts to ensure UI updates during animation
                checkStatus(); // First immediate refresh

                // Stagger additional refreshes to catch the UI during animation
                new android.os.Handler(android.os.Looper.getMainLooper()).postDelayed(() -> {
                    Log.d(TAG, "Secondary refresh during animation");
                    checkStatus();
                }, 100); // 100ms delay

                new android.os.Handler(android.os.Looper.getMainLooper()).postDelayed(() -> {
                    Log.d(TAG, "Third refresh during animation");
                    checkStatus();
                }, 500); // 500ms delay
            }
        };
    }

    private void registerBroadcastReceiver() {
        if (getActivity() != null && transactionSuccessReceiver != null) {
            IntentFilter filter = new IntentFilter("com.example.earthwallet.TRANSACTION_SUCCESS");
            try {
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.TIRAMISU) {
                    requireActivity().getApplicationContext().registerReceiver(transactionSuccessReceiver, filter, Context.RECEIVER_NOT_EXPORTED);
                } else {
                    requireActivity().getApplicationContext().registerReceiver(transactionSuccessReceiver, filter);
                }
                Log.d(TAG, "Registered transaction success receiver");
            } catch (Exception e) {
                Log.e(TAG, "Failed to register broadcast receiver", e);
            }
        }
    }

    private void showLoading(boolean loading) {
        // Use the reusable LoadingOverlay component
        if (loadingOverlay != null) {
            if (loading) {
                loadingOverlay.show();
                if (errorText != null) errorText.setVisibility(View.GONE);
                hideStatusFragments();
            } else {
                loadingOverlay.hide();
            }
        }
    }

    // Debug helpers to surface responses in-app
    private void showAlert(String title, String json) {
        try {
            showJsonDialog(title != null ? title : "Response", json);
        } catch (Exception ignored) {}
    }

    private void showJsonDialog(String title, String json) {
        try {
            new AlertDialog.Builder(getContext())
                    .setTitle(title != null ? title : "Response")
                    .setMessage(json != null ? json : "(empty)")
                    .setPositiveButton("OK", null)
                    .show();
        } catch (Exception ignored) {}
    }

    // Show dialog and run a callback after it's dismissed (used to defer follow-up actions)
    private void showJsonDialogThen(String title, String json, Runnable then) {
        try {
            AlertDialog dialog = new AlertDialog.Builder(getContext())
                    .setTitle(title != null ? title : "Response")
                    .setMessage(json != null ? json : "(empty)")
                    .setPositiveButton("OK", null)
                    .create();
            dialog.setOnDismissListener(d -> {
                try { if (then != null) then.run(); } catch (Exception ignored) {}
            });
            dialog.show();
        } catch (Exception ignored) {}
    }

    private void hideStatusFragments() {
        FragmentManager fm = getChildFragmentManager();
        Fragment current = fm.findFragmentById(R.id.anml_root);
        if (current != null) {
            fm.beginTransaction().remove(current).commit();
        }
    }

    private void showRegisterFragment() {
        hideStatusFragments();
        ANMLRegisterFragment fragment;
        if (currentRegistrationReward != null) {
            fragment = ANMLRegisterFragment.newInstance(currentRegistrationReward);
        } else {
            fragment = ANMLRegisterFragment.newInstance();
        }
        fragment.setANMLRegisterListener(this);
        getChildFragmentManager().beginTransaction()
                .replace(R.id.anml_root, fragment)
                .commit();
    }

    private void showClaimFragment() {
        hideStatusFragments();
        ANMLClaimFragment fragment = ANMLClaimFragment.newInstance();
        fragment.setANMLClaimListener(this);
        getChildFragmentManager().beginTransaction()
                .replace(R.id.anml_root, fragment)
                .commit();
    }

    private void showCompleteFragment() {
        hideStatusFragments();
        ANMLCompleteFragment fragment = ANMLCompleteFragment.newInstance();
        getChildFragmentManager().beginTransaction()
                .replace(R.id.anml_root, fragment)
                .commit();
    }

    @Override
    public void onRegisterRequested() {
        Intent i = new Intent(getContext(), com.example.earthwallet.ui.host.HostActivity.class);
        i.putExtra("fragment_to_show", "camera_mrz_scanner");
        startActivity(i);
    }

    @Override
    public void onClaimRequested() {
        try {
            org.json.JSONObject exec = new org.json.JSONObject();
            exec.put("claim_anml", new org.json.JSONObject());

            Intent ei = new Intent(getContext(), TransactionActivity.class);
            ei.putExtra(TransactionActivity.EXTRA_TRANSACTION_TYPE, TransactionActivity.TYPE_SECRET_EXECUTE);
            ei.putExtra(TransactionActivity.EXTRA_CONTRACT_ADDRESS, Constants.REGISTRATION_CONTRACT);
            ei.putExtra(TransactionActivity.EXTRA_CODE_HASH, Constants.REGISTRATION_HASH);
            ei.putExtra(TransactionActivity.EXTRA_EXECUTE_JSON, exec.toString());
            // Funds/memo/lcd are optional; default LCD is used in the bridge
            startActivityForResult(ei, REQ_EXECUTE);
        } catch (Exception e) {
            Toast.makeText(getContext(), "Failed to start claim: " + e.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }

    // Launch the reusable query activity to check registration/claim status
    private void checkStatus() {
        try {
            showLoading(true);

            // Get wallet address using secure just-in-time mnemonic access
            String address;
            try {
                if (!com.example.earthwallet.wallet.services.SecureWalletManager.isWalletAvailable(getContext())) {
                    showLoading(false);
                    showRegisterFragment();
                    return;
                }
                
                address = com.example.earthwallet.wallet.services.SecureWalletManager.getWalletAddress(getContext());
                if (TextUtils.isEmpty(address)) {
                    showLoading(false);
                    showRegisterFragment();
                    return;
                }
            } catch (Exception e) {
                Log.e(TAG, "Failed to get wallet address: " + e.getMessage());
                showLoading(false);
                showRegisterFragment();
                return;
            }

            // Build query object for the bridge (no proxy)
            JSONObject q = new JSONObject();
            JSONObject inner = new JSONObject();
            inner.put("address", address);
            q.put("query_registration_status", inner);

            // Use SecretQueryService directly in background thread
            new Thread(() -> {
                try {
                    // Check wallet availability without retrieving mnemonic
                    if (!com.example.earthwallet.wallet.services.SecureWalletManager.isWalletAvailable(getContext())) {
                        if (getActivity() != null) {
                            getActivity().runOnUiThread(() -> {
                                showLoading(false);
                                if (errorText != null) {
                                    errorText.setText("No wallet found");
                                    errorText.setVisibility(View.VISIBLE);
                                }
                            });
                        }
                        return;
                    }
                    
                    SecretQueryService queryService = new SecretQueryService(getContext());
                    JSONObject result = queryService.queryContract(
                        Constants.REGISTRATION_CONTRACT,
                        Constants.REGISTRATION_HASH,
                        q
                    );
                    
                    // Format result to match expected format
                    JSONObject response = new JSONObject();
                    response.put("success", true);
                    response.put("result", result);
                    
                    // Handle result on UI thread
                    if (getActivity() != null) {
                        getActivity().runOnUiThread(() -> {
                            handleRegistrationQueryResult(response.toString());
                        });
                    }
                    
                } catch (Exception e) {
                    Log.e(TAG, "Registration status query failed", e);
                    if (getActivity() != null) {
                        getActivity().runOnUiThread(() -> {
                            showLoading(false);
                            if (errorText != null) {
                                errorText.setText("Failed to check status: " + e.getMessage());
                                errorText.setVisibility(View.VISIBLE);
                            }
                        });
                    }
                }
            }).start();
        } catch (Exception e) {
            Log.e(TAG, "checkStatus failed", e);
            showLoading(false);
            if (errorText != null) {
                errorText.setText("Failed to check status: " + e.getMessage());
                errorText.setVisibility(View.VISIBLE);
            }
        }
    }

    private void handleRegistrationQueryResult(String json) {
        showLoading(false);
        boolean wasSuppressed = suppressNextQueryDialog;
        suppressNextQueryDialog = false;
        
        try {
            if (TextUtils.isEmpty(json)) {
                if (errorText != null) {
                    errorText.setText("No response from bridge.");
                    errorText.setVisibility(View.VISIBLE);
                }
                return;
            }
            
            JSONObject root = new JSONObject(json);
            boolean success = root.optBoolean("success", false);
            if (!success) {
                if (errorText != null) {
                    errorText.setText("Bridge returned error.");
                    errorText.setVisibility(View.VISIBLE);
                }
                return;
            }
            JSONObject result = root.optJSONObject("result");
            if (result == null) {
                result = root;
            }

            // Extract registration reward if available
            currentRegistrationReward = result.optString("registration_reward", null);

            if ("no_wallet".equals(result.optString("status", ""))) {
                showRegisterFragment();
                return;
            }

            boolean registered = result.optBoolean("registration_status", false);
            if (!registered) {
                showRegisterFragment();
                return;
            }

            long lastClaim = result.optLong("last_claim", 0L);
            long nextClaimMillis = (lastClaim / 1000000L) + ONE_DAY_MILLIS;
            long now = System.currentTimeMillis();
            if (now > nextClaimMillis) {
                showClaimFragment();
            } else {
                showCompleteFragment();
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to parse bridge result", e);
            if (errorText != null) {
                errorText.setText("Invalid result from bridge.");
                errorText.setVisibility(View.VISIBLE);
            }
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == REQ_EXECUTE) {
            // Hide loading screen that might be showing
            showLoading(false);

            if (resultCode == getActivity().RESULT_OK) {
                Log.d(TAG, "ANML claim transaction succeeded");
                // Transaction successful - navigate optimistically to complete screen
                showCompleteFragment();
                // Also refresh status (broadcast receiver will handle additional refreshes)
                checkStatus();
            } else {
                String error = data != null ? data.getStringExtra(TransactionActivity.EXTRA_ERROR) : "Unknown error";
                Log.e(TAG, "ANML claim transaction failed: " + error);
                // Transaction failed or was cancelled - refresh to ensure UI is correct
                checkStatus();
            }
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        // Cleanup loading overlay
        if (loadingOverlay != null && getContext() != null) {
            loadingOverlay.cleanup(getContext());
        }

        // Unregister broadcast receiver
        if (transactionSuccessReceiver != null && getContext() != null) {
            try {
                requireActivity().getApplicationContext().unregisterReceiver(transactionSuccessReceiver);
                Log.d(TAG, "Unregistered transaction success receiver");
            } catch (IllegalArgumentException e) {
                // Receiver was not registered, ignore
                Log.d(TAG, "Receiver was not registered");
            } catch (Exception e) {
                Log.e(TAG, "Error unregistering receiver", e);
            }
        }
    }
}

File: app/src/main/java/com/example/earthwallet/ui/pages/anml/ANMLClaimMainFragment.kt
package network.erth.wallet.ui.pages.anml

import network.erth.wallet.R
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.lifecycleScope
import network.erth.wallet.Constants
import network.erth.wallet.ui.components.LoadingOverlay
import network.erth.wallet.wallet.services.SecretKClient
import network.erth.wallet.wallet.services.SecureWalletManager
import network.erth.wallet.wallet.services.TransactionExecutor
import kotlinx.coroutines.launch
import org.json.JSONObject

class ANMLClaimMainFragment : Fragment(), ANMLRegisterFragment.ANMLRegisterListener, ANMLClaimFragment.ANMLClaimListener {

    companion object {
        private const val TAG = "ANMLClaimFragment"
        private const val ONE_DAY_MILLIS = 24L * 60L * 60L * 1000L

        @JvmStatic
        fun newInstance(): ANMLClaimMainFragment = ANMLClaimMainFragment()
    }

    private var errorText: TextView? = null
    private var loadingOverlay: LoadingOverlay? = null
    private var fragmentContainer: View? = null

    private var suppressNextQueryDialog = false
    private var currentRegistrationReward: String? = null
    private var transactionSuccessReceiver: BroadcastReceiver? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_anml_claim_main, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        try {
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize SecretWallet", e)
        }

        errorText = view.findViewById(R.id.anml_error_text)
        loadingOverlay = view.findViewById(R.id.loading_overlay)
        fragmentContainer = view.findViewById(R.id.anml_root)

        // Initialize the loading overlay with this fragment for Glide
        loadingOverlay?.initializeWithFragment(this)

        setupBroadcastReceiver()
        registerBroadcastReceiver()

        // Start status check
        checkStatus()
    }

    private fun setupBroadcastReceiver() {
        transactionSuccessReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {

                // Start multiple refresh attempts to ensure UI updates during animation
                checkStatus() // First immediate refresh

                // Stagger additional refreshes to catch the UI during animation
                Handler(Looper.getMainLooper()).postDelayed({
                    checkStatus()
                }, 100) // 100ms delay

                Handler(Looper.getMainLooper()).postDelayed({
                    checkStatus()
                }, 500) // 500ms delay
            }
        }
    }

    private fun registerBroadcastReceiver() {
        if (activity != null && transactionSuccessReceiver != null) {
            val filter = IntentFilter("network.erth.wallet.TRANSACTION_SUCCESS")
            try {
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.TIRAMISU) {
                    requireActivity().applicationContext.registerReceiver(transactionSuccessReceiver, filter, Context.RECEIVER_NOT_EXPORTED)
                } else {
                    requireActivity().applicationContext.registerReceiver(transactionSuccessReceiver, filter)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to register broadcast receiver", e)
            }
        }
    }

    private fun showLoading(loading: Boolean) {
        // Use the reusable LoadingOverlay component
        loadingOverlay?.let { overlay ->
            if (loading) {
                overlay.show()
                errorText?.visibility = View.GONE
                hideStatusFragments()
            } else {
                overlay.hide()
            }
        }
    }

    // Debug helpers to surface responses in-app
    private fun showAlert(title: String?, json: String?) {
        try {
            showJsonDialog(title ?: "Response", json)
        } catch (ignored: Exception) {}
    }

    private fun showJsonDialog(title: String?, json: String?) {
        try {
            AlertDialog.Builder(requireContext())
                .setTitle(title ?: "Response")
                .setMessage(json ?: "(empty)")
                .setPositiveButton("OK", null)
                .show()
        } catch (ignored: Exception) {}
    }

    // Show dialog and run a callback after it's dismissed (used to defer follow-up actions)
    private fun showJsonDialogThen(title: String?, json: String?, then: Runnable?) {
        try {
            val dialog = AlertDialog.Builder(requireContext())
                .setTitle(title ?: "Response")
                .setMessage(json ?: "(empty)")
                .setPositiveButton("OK", null)
                .create()
            dialog.setOnDismissListener {
                try {
                    then?.run()
                } catch (ignored: Exception) {}
            }
            dialog.show()
        } catch (ignored: Exception) {}
    }

    private fun hideStatusFragments() {
        val fm = childFragmentManager
        val current = fm.findFragmentById(R.id.anml_root)
        if (current != null) {
            fm.beginTransaction().remove(current).commit()
        }
    }

    private fun showRegisterFragment() {
        hideStatusFragments()
        val fragment = if (currentRegistrationReward != null) {
            ANMLRegisterFragment.newInstance(currentRegistrationReward!!)
        } else {
            ANMLRegisterFragment.newInstance()
        }
        fragment.setANMLRegisterListener(this)
        childFragmentManager.beginTransaction()
            .replace(R.id.anml_root, fragment)
            .commit()
    }

    private fun showClaimFragment() {
        hideStatusFragments()
        val fragment = ANMLClaimFragment.newInstance()
        fragment.setANMLClaimListener(this)
        childFragmentManager.beginTransaction()
            .replace(R.id.anml_root, fragment)
            .commit()
    }

    private fun showCompleteFragment() {
        hideStatusFragments()
        val fragment = ANMLCompleteFragment.newInstance()
        childFragmentManager.beginTransaction()
            .replace(R.id.anml_root, fragment)
            .commit()
    }

    override fun onRegisterRequested() {
        val i = Intent(context, network.erth.wallet.ui.host.HostActivity::class.java)
        i.putExtra("fragment_to_show", "camera_mrz_scanner")
        startActivity(i)
    }

    override fun onClaimRequested() {
        lifecycleScope.launch {
            try {
                val exec = JSONObject()
                exec.put("claim_anml", JSONObject())

                val result = TransactionExecutor.executeContract(
                    fragment = this@ANMLClaimMainFragment,
                    contractAddress = Constants.REGISTRATION_CONTRACT,
                    message = exec,
                    codeHash = Constants.REGISTRATION_HASH,
                    gasLimit = 350_000,  // Needs ~302k gas, set to 350k for safety
                    contractLabel = "Registration Contract:"
                )

                result.onSuccess {
                    // Transaction successful - navigate to complete screen
                    showCompleteFragment()
                    // Broadcast receiver will handle status refreshes
                }.onFailure { error ->
                    if (error.message != "Transaction cancelled by user" &&
                        error.message != "Authentication failed") {
                        Log.e(TAG, "ANML claim failed: ${error.message}")
                        Toast.makeText(context, "Claim failed: ${error.message}", Toast.LENGTH_LONG).show()
                    }
                    // Transaction failed or was cancelled - refresh to ensure UI is correct
                    checkStatus()
                }

            } catch (e: Exception) {
                Toast.makeText(context, "Failed: ${e.message}", Toast.LENGTH_SHORT).show()
            }
        }
    }

    // Launch the reusable query activity to check registration/claim status
    private fun checkStatus() {
        try {
            showLoading(true)

            // Get wallet address using SecureWalletManager
            val address: String
            try {
                if (!SecureWalletManager.isWalletAvailable(requireContext())) {
                    showLoading(false)
                    showRegisterFragment()
                    return
                }

                address = SecureWalletManager.getWalletAddress(requireContext()) ?: ""
                if (TextUtils.isEmpty(address)) {
                    showLoading(false)
                    showRegisterFragment()
                    return
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to get wallet address: ${e.message}")
                showLoading(false)
                showRegisterFragment()
                return
            }

            // Build query object for the bridge (no proxy)
            val q = JSONObject()
            val inner = JSONObject()
            inner.put("address", address)
            q.put("query_registration_status", inner)

            // Use SecretKClient directly
            lifecycleScope.launch {
                try {
                    // Check wallet availability without retrieving mnemonic
                    if (!SecureWalletManager.isWalletAvailable(requireContext())) {
                        showLoading(false)
                        errorText?.let { errorText ->
                            errorText.text = "No wallet found"
                            errorText.visibility = View.VISIBLE
                        }
                        return@launch
                    }

                    val result = SecretKClient.queryContractJson(
                        Constants.REGISTRATION_CONTRACT,
                        q,
                        Constants.REGISTRATION_HASH
                    )

                    // Extract data from wrapper
                    val actualResult = if (result.has("data")) {
                        result.getJSONObject("data")
                    } else {
                        result
                    }

                    // Format result to match expected format
                    val response = JSONObject()
                    response.put("success", true)
                    response.put("result", actualResult)

                    // Handle result
                    handleRegistrationQueryResult(response.toString())

                } catch (e: Exception) {
                    Log.e(TAG, "Registration status query failed", e)
                    val fullError = "Failed to check status:\n${e.javaClass.simpleName}: ${e.message}\n${e.stackTrace.take(3).joinToString("\n")}"
                    showLoading(false)
                    errorText?.let { errorText ->
                        errorText.text = fullError
                        errorText.visibility = View.VISIBLE
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "checkStatus failed", e)
            showLoading(false)
            errorText?.let { errorText ->
                errorText.text = "Error: ${e.message}"
                errorText.visibility = View.VISIBLE
            }
        }
    }

    private fun handleRegistrationQueryResult(json: String) {
        showLoading(false)
        val wasSuppressed = suppressNextQueryDialog
        suppressNextQueryDialog = false

        try {
            if (TextUtils.isEmpty(json)) {
                errorText?.let { errorText ->
                    errorText.text = "No response from bridge."
                    errorText.visibility = View.VISIBLE
                }
                return
            }

            val root = JSONObject(json)
            val success = root.optBoolean("success", false)
            if (!success) {
                errorText?.let { errorText ->
                    errorText.text = "Bridge returned error."
                    errorText.visibility = View.VISIBLE
                }
                return
            }
            val result = root.optJSONObject("result") ?: root

            // Extract registration reward if available
            currentRegistrationReward = result.optString("registration_reward", null).takeIf { it != "null" }

            if ("no_wallet" == result.optString("status", "")) {
                showRegisterFragment()
                return
            }

            val registered = result.optBoolean("registration_status", false)
            if (!registered) {
                showRegisterFragment()
                return
            }

            val lastClaim = result.optLong("last_claim", 0L)
            val nextClaimMillis = (lastClaim / 1000000L) + ONE_DAY_MILLIS
            val now = System.currentTimeMillis()
            if (now > nextClaimMillis) {
                showClaimFragment()
            } else {
                showCompleteFragment()
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse bridge result", e)
            errorText?.let { errorText ->
                errorText.text = "Invalid result from bridge."
                errorText.visibility = View.VISIBLE
            }
        }
    }


    override fun onDestroy() {
        super.onDestroy()

        // Cleanup loading overlay
        if (loadingOverlay != null && context != null) {
            loadingOverlay!!.cleanup(context)
        }

        // Unregister broadcast receiver
        if (transactionSuccessReceiver != null && context != null) {
            try {
                requireActivity().applicationContext.unregisterReceiver(transactionSuccessReceiver)
            } catch (e: IllegalArgumentException) {
                // Receiver was not registered, ignore
            } catch (e: Exception) {
                Log.e(TAG, "Error unregistering receiver", e)
            }
        }
    }
}